File: src/bitmask.c
#include "bitmask.h"
#include "rte_malloc.h"
#include "debug.h"

struct mg_bitmask * mg_bitmask_create(uint16_t size){
  uint16_t n_blocks = (size-1)/64 + 1;
  struct mg_bitmask *mask = rte_zmalloc(NULL, sizeof(struct mg_bitmask) + (size-1)/64 * 8 + 8, 0);
  mask->size = size;
  mask->n_blocks = n_blocks;
  return mask;
}
void mg_bitmask_free(struct mg_bitmask * mask){
  rte_free(mask);
}

void mg_bitmask_clear_all(struct mg_bitmask * mask){
  // TODO: check if memset() would be faster for 64bit values...
  uint16_t i;
  for(i=0; i< mask->n_blocks; i++){
    mask->mask[i] = 0;
  }
}

// This will only touch the first n bits. If other bits are set/cleared, they
// will not be affected
void mg_bitmask_set_n_one(struct mg_bitmask * mask, uint16_t n){
  // TODO: check if memset() would be faster for 64bit values...
  uint64_t * msk = mask->mask;
  while(n>=64){
    *msk = 0xffffffffffffffff;
    n -= 64;
    msk++;
  }
  if(n & 0x3f){
    *msk |= (0xffffffffffffffff >> (64-n));
  }
}

void mg_bitmask_set_all_one(struct mg_bitmask * mask){
  // TODO: check if memset() would be faster for 64bit values...
  // TODO: use mg_bitmask_set_n_one instead...
  uint16_t i;
  for(i=0; i< mask->n_blocks; i++){
    mask->mask[i] = 0xffffffffffffffff;
  }
  // FIXME XXX TODO why do we need this??? can't we just set all blocks to 1s?
  if(mask->size & 0x3f){
    mask->mask[mask->n_blocks-1] = (0xffffffffffffffff >> (64-(mask->size & 0x3f)));
  }
}

uint8_t mg_bitmask_get_bit(struct mg_bitmask * mask, uint16_t n){
  // printf("CCC get bit %d\n", n);
  // printhex("mask = ", mask, 30);
  // printhex("mask = ", mask->mask, 30);
  // uint64_t r1 = mask->mask[n/64] & (1ULL<< (n&0x3f));
  // printhex("r1 = ", &r1, 8);
  uint8_t result = ( (mask->mask[n/64] & (1ULL<< (n&0x3f))) != 0);
  // printf("result = %d\n", (int)result);
  return result;
}

void mg_bitmask_set_bit(struct mg_bitmask * mask, uint16_t n){
  //printf(" CC set bit nr %d\n", n);
  //printhex("mask = ", mask->mask, 8*3);
  mask->mask[n/64] |= (1ULL<< (n&0x3f));
  //printhex("mask = ", mask->mask, 8*3);
}

void mg_bitmask_clear_bit(struct mg_bitmask * mask, uint16_t n){
  mask->mask[n/64] &= ~(1ULL<< (n&0x3f));
}

void mg_bitmask_and(struct mg_bitmask * mask1, struct mg_bitmask * mask2, struct mg_bitmask * result){
  uint16_t i;
  for(i=0; i< mask1->n_blocks; i++){
    result->mask[i] = mask1->mask[i] & mask2->mask[i];
  }
}

void mg_bitmask_xor(struct mg_bitmask * mask1, struct mg_bitmask * mask2, struct mg_bitmask * result){
  uint16_t i;
  for(i=0; i< mask1->n_blocks; i++){
    result->mask[i] = mask1->mask[i] ^ mask2->mask[i];
  }
}

void mg_bitmask_or(struct mg_bitmask * mask1, struct mg_bitmask * mask2, struct mg_bitmask * result){
  uint16_t i;
  for(i=0; i< mask1->n_blocks; i++){
    result->mask[i] = mask1->mask[i] | mask2->mask[i];
  }
}

void mg_bitmask_not(struct mg_bitmask * mask1, struct mg_bitmask * result){
  uint16_t i;
  for(i=0; i< mask1->n_blocks; i++){
    result->mask[i] = ~(mask1->mask[i]);
  }
}

File: lua/include/bitmask.lua
---------------------------------
--- @file bitmask.lua
--- @brief Bitmask ...
--- @todo TODO docu
---------------------------------

local log = require "log"
local ffi = require "ffi"
ffi.cdef [[
struct mg_bitmask{
  uint16_t size;
  uint16_t n_blocks;
  uint64_t mask[0];
};
struct mg_bitmask * mg_bitmask_create(uint16_t size);
void mg_bitmask_free(struct mg_bitmask * mask);
void mg_bitmask_set_n_one(struct mg_bitmask * mask, uint16_t n);
void mg_bitmask_set_all_one(struct mg_bitmask * mask);
void mg_bitmask_clear_all(struct mg_bitmask * mask);
uint8_t mg_bitmask_get_bit(struct mg_bitmask * mask, uint16_t n);
void mg_bitmask_set_bit(struct mg_bitmask * mask, uint16_t n);
void mg_bitmask_clear_bit(struct mg_bitmask * mask, uint16_t n);
void mg_bitmask_and(struct mg_bitmask * mask1, struct mg_bitmask * mask2, struct mg_bitmask * result);
void mg_bitmask_xor(struct mg_bitmask * mask1, struct mg_bitmask * mask2, struct mg_bitmask * result);
void mg_bitmask_or(struct mg_bitmask * mask1, struct mg_bitmask * mask2, struct mg_bitmask * result);
void mg_bitmask_not(struct mg_bitmask * mask1, struct mg_bitmask * result);
]]


mod = {}

local mg_bitMask = {}
--mg_bitMask.__index = mg_bitMask

--- Create a Bitmask
--- The mask is internally built from blocks of 64bit integers. Hence a Bitmask
--- of a size <<64 yields significant overhead
--- @param size Size of the bitmask in number of bits
--- @return Wrapper table around the bitmask
function mod.createBitMask(size)
  return setmetatable({
    bitmask = ffi.gc(ffi.C.mg_bitmask_create(size), function (self)
      log:debug("I HAVE BEEN DESTRUCTED")
      ffi.C.mg_bitmask_free(self)
    end )
  }, mg_bitMask)
end

---
--- @param bitmasks
--- @todo TODO: think of a better solution - meh
function mod.linkToArray(bitmasks)
  array = ffi.new("struct mg_bitmask*[?]", #bitmasks)
  local i = 0
  for _,m in pairs(bitmasks) do
    array[i] = m.bitmask
    i = i + 1
  end
  return array
end

--- sets the first n bits in a bitmask to 1
--- other bits remain unchanged
--- @param n
function mg_bitMask:setN(n)
  ffi.C.mg_bitmask_set_n_one(self.bitmask, n)
  return self
end

--- sets all bits in a bitmask to 0
function mg_bitMask:clearAll()
  ffi.C.mg_bitmask_clear_all(self.bitmask)
  return self
end

--- sets all bits in a bitmask to 1
function mg_bitMask:setAll()
  ffi.C.mg_bitmask_set_all_one(self.bitmask)
  return self
end

--- Index metamethod for mg_bitMask
--- @param x Bit index. Index starts at 1 according to the LUA standard (1 indexes the first bit in the bitmask)
--- @return For numeric indices: true, when corresponding bit is 1, false otherwise.
function mg_bitMask:__index(x)
  -- access
  --print(" bit access")
  --print(" type " .. type(x))
  --print(" x = " .. tostring(x))
  if(type(x) == "number") then
    return (ffi.C.mg_bitmask_get_bit(self.bitmask, x - 1) ~= 0)
  else
    return mg_bitMask[x]
  end
end

--- Newindex metamethod for mg_bitMask
--- @param x Bit index. Index starts at 1 according to the LUA standard (1 indexes the first bit in the bitmask)
--- @param y Assigned value to the index (bit is cleared for y==0 and set otherwise)
function mg_bitMask:__newindex(x, y)
  --print ("new index")
  if(y == 0) then
    -- clear bit
    --print("clear")
    return ffi.C.mg_bitmask_clear_bit(self.bitmask, x - 1)
  else
    -- set bit
    --print("set")
    return ffi.C.mg_bitmask_set_bit(self.bitmask, x - 1)
  end
end

do
	local function it(self, i)
		if i >= self.bitmask.size then
			return nil
		end
		return i + 1, self[i+1]
	end

	function mg_bitMask.__ipairs(self)
		return it, self, 0
	end
end

--- Bitwise and
--- @param mask1
--- @param mask2
--- @param result
---  result = mask1 band mask2
function mod.band(mask1, mask2, result)
  ffi.C.mg_bitmask_and(mask1.bitmask, mask2.bitmask, result.bitmask)
end

--- Bitwise or
--- @param mask1
--- @param mask2
--- @param result
---  result = mask1 bor mask2
function mod.bor(mask1, mask2, result)
  ffi.C.mg_bitmask_or(mask1.bitmask, mask2.bitmask, result.bitmask)
end

--- Bitwise xor
--- @param mask1
--- @param mask2
--- @param result
---  result = bask1 bxor mask2
function mod.bxor(mask1, mask2, result)
  ffi.C.mg_bitmask_xor(mask1.bitmask, mask2.bitmask, result.bitmask)
end

--- Bitwise not
--- @param mask
--- @param result
---  result = not mask
function mod.bnot(mask, result)
  ffi.C.mg_bitmask_not(mask.bitmask, result.bitmask)
end

return mod
